packages/sheets-ui/src/views/clipboard/ClipboardPopupMenu.tsx

```diff
@@ -0,0 +1,221 @@
+];
+
+const useMenuPosition = (pasteOptionsCache: IPasteOptionCache | null) => {
+    const univerInstanceService = useDependency(IUniverInstanceService);
+    const renderManagerService = useDependency(IRenderManagerService);
+
+    const workbook = useActiveWorkbook();
+
+    const range = pasteOptionsCache?.target.pastedRange;
+    if (!range || !workbook) return null;
+
+    const anchor = {
+        row: range.rows[range.rows.length - 1],
+        col: range.cols[range.cols.length - 1],
+    };
+
+    if (anchor.row < 0 || anchor.col < 0) {
+        return null;
+    }
+
+    const ru = renderManagerService.getRenderById(workbook.getUnitId());
+    const sheetSkeletonManagerService = ru?.with(SheetSkeletonManagerService);
+    const selectionRenderService = ru?.with(ISheetSelectionRenderService);
+
+    const sheetObject = getSheetObject(univerInstanceService, renderManagerService);
+    if (!sheetObject || !selectionRenderService) return null;
+
+    const { scene } = sheetObject;
+    const skeleton = sheetSkeletonManagerService?.getCurrentSkeleton();
+    const viewport = selectionRenderService.getViewPort();
+    const scaleX = scene?.scaleX;
+    const scaleY = scene?.scaleY;
+    const scrollXY = scene?.getViewportScrollXY(viewport);
+    const canvas = scene.getEngine()?.getCanvas();
+    if (!scaleX || !scene || !scaleX || !scaleY || !scrollXY) return null;
+    const x = skeleton?.getNoMergeCellPositionByIndex(anchor.row, anchor.col).endX || 0;
+    const y = skeleton?.getNoMergeCellPositionByIndex(anchor.row, anchor.col).endY || 0;
+    let positionX = skeleton?.convertTransformToOffsetX(x, scaleX, scrollXY) ?? -9999;
+    let positionY = skeleton?.convertTransformToOffsetY(y, scaleY, scrollXY) ?? -9999;
+
+    const canvasWidth = canvas?.getWidth();
+    const canvasHeight = canvas?.getHeight();
+
+    if (!canvasWidth || !canvasHeight) return null;
+
+    if (positionX + 50 > canvasWidth) {
+        positionX = canvasWidth - 100;
+    }
+
+    if (positionY + 50 > canvasHeight) {
+        positionY = canvasHeight - 100;
+    }
+
+    return {
+        positionX,
+        positionY,
+    };
+};
+
+export const ClipboardPopupMenu = () => {
+    const clipboardService = useDependency(ISheetClipboardService);
+    const showMenu = useObservable(clipboardService.showMenu$, true);
+    const pasteOptionsCache = useObservable(clipboardService.pasteOptionsCache$, null);
+    const renderManagerService = useDependency(IRenderManagerService);
+    const localeService = useDependency(LocaleService);
+    const commandService = useDependency(ICommandService);
+
+    const [menuHovered, setMenuHovered] = useState(false);
+    const [menuVisible, setMenuVisible] = useState(false);
+
+    const [_, setVersion] = useState(Math.random());
+
+    const relativePosition = useMenuPosition(pasteOptionsCache);
+
+    const workbook = useActiveWorkbook();
+    const { sheetSkeletonManagerService } = useMemo(() => {
```

Comment:
If it is not necessary, do not pass values in an object, which will give GC more jobs to do.